pkg/spanfilter/spanfilter.go

```diff
@@ -0,0 +1,238 @@
+}
+
+type filterPolicy struct {
+	Include *splitPolicy
+	Exclude *splitPolicy
+}
+
+// SplitPolicy is the result of parsing a policy from the config file to be
+// specific about the area the given policy is applied to.
+type splitPolicy struct {
+	ResourceMatch  *config.PolicyMatch
+	SpanMatch      *config.PolicyMatch
+	IntrinsicMatch *config.PolicyMatch
+}
+
+func NewSpanFilter(filterPolicies []config.FilterPolicy) (*SpanFilter, error) {
+	var policies []*filterPolicy
+
+	var err error
+	for _, policy := range filterPolicies {
+		err = config.ValidateFilterPolicy(policy)
+		if err != nil {
+			return nil, err
+		}
+
+		p := &filterPolicy{
+			Include: getSplitPolicy(policy.Include),
+			Exclude: getSplitPolicy(policy.Exclude),
+		}
+
+		if p.Include != nil || p.Exclude != nil {
+			policies = append(policies, p)
+		}
+	}
+
+	return &SpanFilter{
+		filterPolicies: policies,
+	}, nil
+}
+
+// applyFilterPolicy returns true if the span should be included in the metrics.
+func (f *SpanFilter) ApplyFilterPolicy(rs *v1.Resource, span *v1_trace.Span) bool {
+	// With no filter policies specified, all spans are included.
+	if len(f.filterPolicies) == 0 {
+		return true
+	}
+
+	for _, policy := range f.filterPolicies {
+		if policy.Include != nil {
+			if !policyMatch(policy.Include, rs, span) {
+				return false
+			}
+		}
+
+		if policy.Exclude != nil {
+			if policyMatch(policy.Exclude, rs, span) {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func stringMatch(matchType config.MatchType, s, pattern string) bool {
+	switch matchType {
+	case config.Strict:
+		return s == pattern
+	case config.Regex:
+		re := regexp.MustCompile(pattern)
+		return re.MatchString(s)
+	default:
+		return false
+	}
+}
+
+// policyMatch returns true when the resource attribtues and span attributes match the policy.
+func policyMatch(policy *splitPolicy, rs *v1.Resource, span *v1_trace.Span) bool {
+	return policyMatchAttrs(policy.ResourceMatch, rs.Attributes) &&
+		policyMatchAttrs(policy.SpanMatch, span.Attributes) &&
+		policyMatchIntrinsicAttrs(policy.IntrinsicMatch, span)
+}
+
+// policyMatchIntrinsicAttrs returns true when all intrinsic values in the policy match the span.
+func policyMatchIntrinsicAttrs(policy *config.PolicyMatch, span *v1_trace.Span) bool {
+	matches := 0
+
+	var attr traceql.Attribute
+	for _, pa := range policy.Attributes {
+		attr = traceql.MustParseIdentifier(pa.Key)
+		switch attr.Intrinsic {
+		// case traceql.IntrinsicDuration:
+		// case traceql.IntrinsicChildCount:
+		// case traceql.IntrinsicParent:
+		case traceql.IntrinsicName:
+			if !stringMatch(policy.MatchType, span.GetName(), pa.Value.(string)) {
+				return false
+			}
+			matches++
+		case traceql.IntrinsicStatus:
+			if !stringMatch(policy.MatchType, span.GetStatus().GetCode().String(), pa.Value.(string)) {
+				return false
+			}
+			matches++
+		case traceql.IntrinsicKind:
+			if !stringMatch(policy.MatchType, span.GetKind().String(), pa.Value.(string)) {
```

Comment:
anyway we can do int matches here and on status? would be much faster

Reply:
Probably yes, but it might be a little clumsy, since we'd need to take this as an int in the config I think.  There is a special yaml struct tag we could use iirc to parse a string as an int, or perhaps some custom unmarshaling.  For now I'm inclined to leave it and come back to it later.

Reply:
I think I can work this.  I'll take a closer look next week.

Reply:
I've added a commit for this.  It made some parts less readable, but seems okay to me.  How's that look?